src/v2/kuadrant/pipeline/tasks/ratelimit.rs

```diff
@@ -1,36 +1,766 @@
+                    match data_item.value.eval(ctx) {
+                        Ok(AttributeState::Available(val)) => match data_item.key.as_str() {
+                            "ratelimit.domain" => {
+                                if let Value::String(s) = val {
+                                    if s.is_empty() {
+                                        return Err(AttributeError::Parse(
+                                            "ratelimit.domain cannot be empty".to_string(),
+                                        ));
+                                    }
+                                    domain = s.to_string();
+                                } else {
+                                    return Err(AttributeError::Parse(
+                                        "ratelimit.domain must be string".to_string(),
+                                    ));
+                                }
+                            }
+                            "ratelimit.hits_addend" => match val {
+                                Value::Int(i) if i >= 0 && i <= u32::MAX as i64 => {
+                                    hits_addend = i as u32;
+                                }
+                                Value::UInt(u) if u <= u32::MAX as u64 => {
+                                    hits_addend = u as u32;
+                                }
+                                _ => {
+                                    return Err(AttributeError::Parse(
+                                        "ratelimit.hits_addend must be 0 <= X <= u32::MAX"
+                                            .to_string(),
+                                    ));
+                                }
+                            },
+                            _ => {}
+                        },
+                        Ok(AttributeState::Pending) => {
+                            return Err(AttributeError::NotAvailable(format!(
+                                "Attribute {} is pending",
+                                data_item.key
+                            )));
+                        }
+                        Err(cel_err) => {
+                            return Err(AttributeError::Retrieval(format!(
+                                "CEL evaluation error: {}",
+                                cel_err
+                            )));
+                        }
+                    }
+                }
+            }
+        }
+        Ok((hits_addend, domain))
+    }
+
+    /// Builds rate limit entries needed for rate limit descriptors
+    fn build_entries(
+        &self,
+        conditional_data: &ConditionalData,
+        ctx: &ReqRespCtx,
+    ) -> Result<Vec<Entry>, EvaluationError> {
+        if conditional_data.predicates.is_empty()
+            || conditional_data.predicates.iter().any(|pred| {
+                pred.test(ctx)
+                    .is_ok_and(|v| matches!(v, AttributeState::Available(true)))
+            })
```

Comment:
I thought `all` the predicates had to be true not `any` to add these data items?

Reply:
The way I thought it worked was this:
- Top-level predicates must all be met to be able to apply the action at all
- For each conditional data item; IF all of it's predicates apply
  - Add the conditional data entry

I thought that's what this did at least here https://github.com/Kuadrant/wasm-shim/blob/main/src/ratelimit_action.rs#L92-L106 (no idea why the old code isn't using `predicates.apply()` there though anyway)